### PR TITLE
update hc::printf

### PR DESCRIPTION
**Pull request by scchan**
- making hc::printf an opt-in feature; user will have to compile with  `-DHCC_ENABLE_ACCELERATOR_PRINTF` to enable it
- auto-initialization and deallocation of printf buffer
- use a global printf buffer; remove explicitly passing the printf buffer to printf as parameter
- process the printf buffer on the host when completion_future::wait() is called or before the printf buffer is deallocated

### CODE REVIEW

File: include/hc_printf.hpp [hunk 1]
```diff
@@ -286,6 +336,7 @@ static inline void processPrintfPackets(PrintfPacket* packets, const unsigned in
     formatString = std::regex_replace(formatString,doubleAmpersandPattern,"%");
     std::printf("%s",formatString.c_str());
   }
+  std::flush(std::cout);
 }
 
 static inline void processPrintfBuffer(PrintfPacket* gpuBuffer) {
```
Comment (whchung) [inline include/hc_printf.hpp:342]: @scchan if i understand the logic correctly, `processPrintBuffer` would still be invoked with a non-`nullptr` `gpuBuffer`. would it be possible to further optimize this?

**Review by whchung — CHANGES_REQUESTED** [review]
@scchan with the new flag now the feature is optional, which would definitely speed up kernels when `printf()` is not used, but it seems print buffer flushing would still take place on the host side.  is it possible to also optimize this?

File: include/hc_printf.hpp (hunk 1 above)
Comment (scchan) [reply include/hc_printf.hpp:342]: If you refer to the following line, the host would first check whether the printf buffer contains any new data before proceeding to do further processing.  If kernels in the applicaton doesn't print anything the overhead should be minimal

https://github.com/RadeonOpenCompute/hcc/pull/498/files#diff-531a057446b7d96d716399f005b31ad8R349

Comment (whchung) [reply include/hc_printf.hpp:342]: got it. with the revised logic it's more clear.

**Review by whchung — APPROVED** [review]

**Pull request merged**